composer/trainer/trainer.py

```diff
@@ -2207,8 +2207,10 @@ def _train_microbatches(self,
                     except TypeError:
                         optimizer.zero_grad()
 
-            # tracker for gradient accumulation
+            # Tracker for gradient accumulation
             current_batch_size = sum([self._train_data_spec.get_num_samples_in_batch(batch) for batch in microbatches])
+            # Cache batch, which will be overwritten by microbatches. Restore after microbatches complete
+            current_batch = self.state.batch
```

Comment:
does this need to be a deepcopy for this to work in case something modifies the batch during `_train_microbatch`?

Reply:
I don't think anything should modify it inside `_train_microbatch`. Maybe an algorithm changes it? but then we probably want the changed version (eg if someone is using a subset of samples?)

I would like to avoid deepcopy for performance reasons -- already seeing overhead from keeping batch on device in diffusion work

Reply:
yeah i understand the performance implications. and yeah, an algorithm could change the batch. it would certainly be possible to write an algorithm that messed up this batch caching...but i guess thats ok